hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -2806,14 +2809,40 @@ private TableGroup consumeAttributeJoin(
 		);
 
 		if ( pathSource instanceof PluralPersistentAttribute ) {
+
 			assert modelPart instanceof PluralAttributeMapping;
 
 			final PluralAttributeMapping pluralAttributeMapping = (PluralAttributeMapping) modelPart;
 
+			if ( pluralAttributeMapping.getMappedType()
+					.getCollectionSemantics()
+					.getCollectionClassification() != CollectionClassification.BAG ) {
+				containsCardinalityAlteringJoin = true;
+			}
+
```

Comment:
I think it would be easiest if you just call the `alterResultCardinality` method and encapsulate some logic there.

```suggestion
			alterResultCardinality( sqmJoin );
```